docs/cudf/source/developer_guide/testing.md

```diff
@@ -0,0 +1,105 @@
+# Testing cuDF
+
+## Tooling
+Tests in cuDF are written using [`pytest`](https://docs.pytest.org/en/latest/).
+Test coverage is measured using [`coverage.py`](https://coverage.readthedocs.io/en/6.4.1/),
+specifically the [`pytest-cov`](https://github.com/pytest-dev/pytest-cov) plugin.
+Code coverage reports are uploaded to [Codecov](https://app.codecov.io/gh/rapidsai/cudf).
+Each PR also indicates whether it increases or decreases in test coverage.
+
+## Test organization
+
+How tests are organized depends on which of the following two groups they fall into:
+
+1. Methods of classes like `DataFrame` or `Series`.
+2. Free functions operating on the above classes like `cudf.merge`.
+
+The former should be organized into directories named for the class (all lowercase), e.g. `dataframe/`.
+
+**Open question**: How should we group functionality within a class? How should we organize free functions? Some options:
```

Comment:
How about using the same grouping we use for the API docs? E.g., for `Series`, we have `test_constructions`, `test_attributes`, `test_conversion`, `test_indexing`, etc.,

https://docs.rapids.ai/api/cudf/nightly/api_docs/series.html

Reply:
I didn't read your options... Yes, I like option (1) the best.

Reply:
@mroeschke raises a good point about trying to use a layout that can in some way be linted. It seems in principle possible to add a linter that would accomplish this by parsing the rst files. It feels like overkill and an overengineered solution in this case, but maybe not? IMO there is a _lot_ of value in tests (and benchmarks, and source code) being sufficiently well organized as to be discoverable.

Reply:
Yeah the overengineering (and maintenance) of linting rules can be problematic, but if there's simple principles that can be enforced by linting, e.g. `docs/**` matches `tests/**` structure, all the better.

I will say option 2 probably will help discoverability and natural organization for (new) contributors. I have noticed in pandas it is difficult for users to know where to put tests when the structure namings contains a lot of "concepts" e.g. `test/frame/indexing/test_indexing.py` vs `test/frame/test_loc.py`

Reply:
I opened #11474 to discuss this further, so we can continue there if we need to go further in terms of the enforcement side of things, but as far as convention I think we have largely settled on matching the docs.